desc/integrals/singularities.py

```diff
@@ -433,91 +558,109 @@ def eval_pt_loop(i, fj):
     # but then need to convert vectors back to rpz
     if kernel.ndim == 3:
         f = xyz2rpz_vec(f, phi=eval_data["phi"])
+
     return f
 
 
-def _singular_part(
-    eval_data,
-    eval_grid,
-    source_data,
-    source_grid,
-    s,
-    q,
-    kernel,
-    interpolator,
-    loop=False,
-):
+def _singular_part(eval_data, source_data, kernel, interpolator, loop=False):
     """Integrate singular point by interpolating to polar grid.
 
     Generally follows sec 3.2.2 of [2], with the following differences:
 
-    - hyperparameter M replaced by s
+    - hyperparameter M replaced by ``st`` and ``sz``.
     - density sigma / function f is absorbed into kernel.
     """
-    source_dtheta = source_grid.spacing[:, 1]
-    source_dzeta = source_grid.spacing[:, 2] / source_grid.NFP
+    eval_grid = interpolator._eval_grid
     eval_theta = jnp.asarray(eval_grid.nodes[:, 1])
     eval_zeta = jnp.asarray(eval_grid.nodes[:, 2])
 
-    r, w, dr, dw = _get_quadrature_nodes(q)
-
-    h_t = jnp.mean(source_dtheta)
-    h_z = jnp.mean(source_dzeta)
-
-    eta = _chi(r)
+    r, w, dr, dw = _get_quadrature_nodes(interpolator.q)
+    r = jnp.abs(r)
     # integrand of eq 38 in [2] except stuff that needs to be interpolated
-    v = eta * s**2 * h_t * h_z / 4 * abs(r) * dr * dw
-    keys = list(set(["|e_theta x e_zeta|"] + kernel.keys))
+    v = (
+        _chi(r)
+        * (interpolator.ht * interpolator.hz)
+        * (interpolator.st * interpolator.sz / 4)
+        * r
+        * dr
+        * dw
+    )
+
+    keys = set(["|e_theta x e_zeta|"] + kernel.keys)
     if "phi" in keys:
-        keys += ["omega"]
+        keys.remove("phi")
+        keys.add("omega")
+    keys = list(keys)
     fsource = [source_data[key] for key in keys]
+    # Note that it is necessary to take the Fourier transforms of the
+    # vector components of the orthonormal polar basis vectors R̂, ϕ̂, Ẑ.
+    # Vector components of the Cartesian basis are not NFP periodic.
+    if isinstance(interpolator, DFTInterpolator):
+        fsource = [interpolator.fourier(val) for val in fsource]
```

Comment:
did you test if having this outside the loop/vmap is actually necessary? f is fixed, so if you vmap over `i` I'm pretty sure jax shouldn't duplicate doing the fft.

Reply:
In general im not a fan of having this logic that depends on which method of interpolation is being used. If you want to keep it we should make a method of both that gets called in either case, and the fft method just returns f as is.

Reply:
No, I tried this way first. I don't like to leave potentially large optimizations up to a JIT compiler. We saved 16gb of memory by assigning a vandermonde matrix to a local variable in the bounce integrals.

Reply:
ok, can we at least do my second comment, and always do this but for FFTInterpolator its a no-op so we can have a cleaner API

Reply:
Yes of course